File: output/deepseek_r1/avatar/direct_translation/python/java/successful_translation/codeforces_350_B.java
import java.util.*;

public class codeforces_350_B {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        sc.nextLine();

        String[] tStr = sc.nextLine().split(" ");
        String[] aStr = sc.nextLine().split(" ");

        int[] t = new int[n + 1];
        int[] a = new int[n + 1];
        t[0] = 0;
        a[0] = 0;
        for (int i = 1; i <= n; i++) {
            t[i] = Integer.parseInt(tStr[i - 1]);
            a[i] = Integer.parseInt(aStr[i - 1]);
        }

        int[] cnt = new int[n + 1];
        for (int i = 0; i <= n; i++) {
            cnt[a[i]]++;
        }

        List<Integer> ans = new ArrayList<>();
        for (int i = 1; i <= n; i++) {
            if (t[i] == 1) {
                List<Integer> crt = new ArrayList<>();
                crt.add(i);
                int x = a[i];
                while (cnt[x] == 1) {
                    crt.add(x);
                    x = a[x];
                }
                if (crt.size() > ans.size()) {
                    ans = new ArrayList<>(crt);
                }
            }
        }

        Collections.reverse(ans);

        System.out.println(ans.size());
        List<String> ansStr = new ArrayList<>();
        for (int num : ans) {
            ansStr.add(Integer.toString(num));
        }
        System.out.println(String.join(" ", ansStr));
    }
}

File: output/llama_4_maverick/avatar/algo_based_translation/java/python/successful_translation/codeforces_350_B.py

def main():
    n = int(input())
    a = [0] * (n + 1)
    b = [0] * (n + 1)
    num = [0] * (n + 1)

    inputs = list(map(int, input().split()))
    for i in range(1, n + 1):
        a[i] = inputs[i - 1]

    inputs = list(map(int, input().split()))
    for i in range(1, n + 1):
        b[i] = inputs[i - 1]

    for i in range(1, n + 1):
        num[b[i]] += 1

    ss = ""
    ans = 0
    st = [False] * (n + 1)

    for i in range(1, n + 1):
        if a[i] == 1:
            st[i] = True
            sb = str(i)
            x = b[i]
            s = 1
            while not st[x] and num[x] == 1 and x != 0 and a[x] != 1:
                sb += " " + str(x)
                st[x] = True
                x = b[x]
                s += 1
            if s > ans:
                ans = s
                ss = ' '.join(sb.split()[::-1])

    print(ans)
    print(ss)

if __name__ == "__main__":
    main()

